### ISSUE TITLE
Facilities alert improvements

### ISSUE THREAD

**Issue by briand44** [uiowa/uiowa#6880]
- Update Dispatch template to use branded template https://github.com/uiowa/uiowa/issues/6880#issuecomment-1724070014
- Add project number at the bottom of the template
- Don't output labels in the template for the following fields:
    - Don't show Traffic Control if value is No
    - Don't show Affected Bus Routes if no routes checked
    - If no building selected don't show in template
    - We should check to see if this is consistent with how we are outputting on an alert node
- Require the category field on the Facilities alert content type split
- Spell check feature is a very nice to have feature
    - Chrome seemed to spell check text fields but maybe other browsers only work on text area fields?

<details>

> The email portion is more straight-forward than our email (text wise) and thinking we can pare down our language and use a different font if that is possible.
> 
> > Here's the P&T template per my other email. I like a couple of things about this, but not an exact match to what we need. We can discuss in our meeting. Thanks.
> 
> ![Image](https://user-images.githubusercontent.com/1017022/264413113-8cb7a55f-da92-4bec-9bf9-ccfa63325131.png)

> While inspecting what is wired up vs what is sent, I don't see the project number (AlertProjectNumber) included in the email. That is a new field correct? Should it be included?
> 
> Update: Brian suggested this be part of a separate slice: [#6823 (comment)](https://github.com/uiowa/uiowa/issues/6823#issuecomment-1699513676)

</details>

```[tasklist]
### Tasks
- [x] Update Dispatch template to use branded template Facilities alert improvements #6880
- [x] Add project number to bottom of template
- [x] Require category field for Facilities alerts
- [x] Don't output labels in the template for: If no building selected don't show in template https://apps.its.uiowa.edu/dispatch/help/freemarker#ifelse
- [x] Don't output labels in the template for: Don't show Traffic Control if value is No https://apps.its.uiowa.edu/dispatch/help/freemarker#ifelse
- [x] Don't output labels in the template for: Don't show Affected Bus Routes if no routes checked https://apps.its.uiowa.edu/dispatch/help/freemarker#ifelse
- [x] Check to see if this is consistent with how we are outputting on an alert node
- [x] Fix the node title config that should be hidden on the form
- [ ] Spell check for both text fields and text areas
```

**Comment by bspeare** [comment]
Here's what the current template looks like:

<img width="809" alt="Screenshot 2023-09-18 at 12 27 51 PM" src="https://github.com/uiowa/uiowa/assets/1036433/2ee85868-ce46-4533-b97d-770924930be9">

**Comment by bspeare** [comment]
Here's what it would look like in a branded email template:

<img width="1093" alt="Screenshot 2023-09-18 at 12 35 48 PM" src="https://github.com/uiowa/uiowa/assets/1036433/48d64390-57c8-4f4d-91ec-3baab0edee98">

**Comment by bspeare** [comment]
Template syntax link: https://apps.its.uiowa.edu/dispatch/help/freemarker#ifelse

**Comment by bspeare** [comment]
It looks like if/else checks in the template syntax only work for the non-branded email templates.  So it appears that we either need to switch back to the basic text email format or use the branded email that outputs empty headings.   Is there an additional option we should be considering?

**Comment by joewhitsitt** [comment]
>It looks like if/else checks in the template syntax only work for the non-branded email templates.

Is that something we should reach out to the Dispatch team about or just take as-is?

**Comment by briand44** [comment]
>Is that something we should reach out to the Dispatch team about or just take as-is?

Ben said he already did contact the Dispatch team and confirmed this. I've got an email out to Facilities.

**Comment by bspeare** [comment]
I have another email out to Dispatch as well because the if/else checks don't seem to be working in the other templates as well.

**Comment by joewhitsitt** [comment]
Going to see about https://github.com/uiowa/uiowa/pull/6925#issuecomment-1735830750

**Comment by joewhitsitt** [comment]
Just a few comments on the spellchecking since it is still part of this issue.

https://caniuse.com/?search=spellcheck

HTML5 has a spellcheck attribute with fairly good browser support these days (chrome 117 plus). Drupal core is aware of it but has stopped short of adding it to all fields. Instead, the user login form makes sure spellcheck is set to false.

The following code could be added to put the spellcheck attribute on all textfields and textareas. However, how a browser works with it differs. Chrome seems to do "while typing" spell check on textfields but then the dotted red line disappears when focus on the form shifts. Firefox seems to show it after initial focus.

```php
/**
 * Implements hook_preprocess_hook().
 */
function sitenow_preprocess_input(&$variables) {
  $admin_context = \Drupal::service('router.admin_context');
  if ($admin_context->isAdminRoute()) {
    if ($variables['element']['#type'] == 'textfield') {
      $variables['attributes']['spellcheck'] = 'true';
    }
  }
}

/**
 * Implements hook_preprocess_hook().
 */
function sitenow_preprocess_textarea(&$variables) {
  $admin_context = \Drupal::service('router.admin_context');
  if ($admin_context->isAdminRoute()) {
      $variables['attributes']['spellcheck'] = 'true';
  }
}
```

- SiteImprove widget does spell checking but is on render (published) I believe.
- We could also include some JS to suggest certain browser extensions (grammarly).
- There are some contrib solutions out there: https://www.drupal.org/project/spellchecker, https://www.drupal.org/project/spellcheck which uses different libraries (see PSpell below)
- Finally, it looks like our Acquia/PHP has PSpell support which we could validate each field against during form submit...
https://www.php.net/manual/en/function.pspell-check.php

**Comment by joewhitsitt** [comment]
TLDR;

If Facilities' construction managers primarily use Firefox to add alerts, we could use the spellcheck attribute scoped to just the alerts form to provide browser based spell check and they will see the underlines while typing and after shifting focus. Not on load though. Other solutions come up short or are more involved to implement.

**Comment by joewhitsitt** [comment]
Adjusted the PR to not close this issue.

>At EoD I explained how conditionals are still not working (template outside this PR) and it was decided to bring the discoveries of https://github.com/uiowa/uiowa/issues/6880#issuecomment-1738019229 (spell checking) to the customer for further discussion before any solution is implemented.

>Everything else looks good.

**Comment by joewhitsitt** [comment]
From standup, going to reach out to Michael A about the conditional template issues

**Comment by joewhitsitt** [comment]
Michael came back with some changes.

>Placeholders are really meant for people creating communications via the GUI

He provided different template markup but also said to check how the data is being sent. Digging into that now.

**Comment by joewhitsitt** [comment]
Don't know what magic Ben implemented to get things branded, but I suppose that would be the next step after https://github.com/uiowa/uiowa/pull/6934 is merged

**Comment by bspeare** [comment]
Here are the steps to use a branded template:

1. Grabbed the code from https://apps.its.uiowa.edu/dispatch/templates/1238199082
2. Edited the campaign communication details https://apps.its.uiowa.edu/dispatch/communications/1238605612
3. Switched the "Template" from "UI Access & Construction Alert - V3" to "UI 1 row x 1 col + 1 row x 2 cols"
4. Went to `row1_content` and hit the source mode and pasted in the code from the template. 
5. I'm not entirely sure if the conditionals will work here because if you try editing `row1_content` the conditionals are stripped out. 

I have updated the template and I can test it after the next release.

**Comment by joewhitsitt** [comment]
>I'm not entirely sure if the conditionals will work here because if you try editing row1_content the conditionals are stripped out.

Ah. Gotcha. It might be basic un-branded then.

**Comment by bspeare** [comment]
I have tested prod and the branded template with conditionals works.

**Comment by pyrello** [comment]
> I have tested prod and the branded template with conditionals works.

Is the next step for this to switch it to the branded template? Anything else?

**Comment by briand44** [comment]
We discussed these changes with Facilities today and came away with a couple additional asks I will create a follow up issue for.

- They would like us to implement the html5 spellcheck attribute for Facilities alert fields
- They are going to send us the order of fields for the node, email template. It sounded like they just wanted to move the category field higher for both of those.
